osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs

```diff
@@ -1,25 +1,80 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using osu.Game.Beatmaps;
 using osu.Game.Rulesets.Judgements;
-using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Judgements;
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Rulesets.Osu.Scoring
 {
     public partial class OsuHealthProcessor : DrainingHealthProcessor
     {
+        private ComboResult currentComboResult = ComboResult.Perfect;
+
         public OsuHealthProcessor(double drainStartTime, double drainLenience = 0)
             : base(drainStartTime, drainLenience)
         {
         }
 
-        protected override int? GetDensityGroup(HitObject hitObject) => (hitObject as IHasComboInformation)?.ComboIndex;
-
         protected override double GetHealthIncreaseFor(JudgementResult result)
+        {
+            if (IsSimulating)
+                return getHealthIncreaseFor(result);
```

Comment:
So the simulation run that determines the drain rate does ignores the existence of the end combo bonus entirely?

Reply:
Yeah. This is not how stable does it but adding the combo bonus during simulation still makes it harder than it should be. Perhaps it could consider the lowest possible combo result, but let's get some feedback on how this feels the way it is first.